File: leet_code/word_search.py
class Solution:
    moves = [(1, 0), (-1, 0), (0, 1), (0, -1)]
    def findWords(self, board, words):
        result = set()
        self.char_map = dict()
        
        for i in range(len(board)):
            for j in range(len(board[i])):
                c = board[i][j]
                if c not in self.char_map: self.char_map[c] = {(i, j)}
                else: self.char_map[c].add((i, j))
        
        for word in words:
            try:
              all_starts = self.char_map[word[0]]
            except KeyError:
              continue  # letter not in board
            for pos in all_starts:
                (prev_i, prev_j) = pos
                self.char_map[word[0]].remove((prev_i, prev_j))
                found = self.findWordsHelper(word[1:], prev_i, prev_j)
                self.char_map[word[0]].add((prev_i, prev_j))
                if found: result.add(word)
        return list(result)
                
    def findWordsHelper(self, word, i, j):
        if len(word) == 0: return True
        c = word[0]
        if c not in self.char_map: return False
        for di, dj in Solution.moves:
            if (i + di, j + dj) in self.char_map[c]:
                # make move, can't return to prev position
                self.char_map[c].remove((i + di, j + dj))
                guess = self.findWordsHelper(word[1:], i+di, j+dj)
                self.char_map[c].add((i + di, j + dj))
                if guess: return True
        
        return False

board = [
  ["a", "a"]
]
words = ["aaa"]
A = Solution()
print(A.findWords(board, words))

File: leet_code/word_search.java
import java.util.List;
import java_datastructures.Trie;

class Solution {
  private TrieNode root;
  private List<String> res;
  public List<String> findWords(char[][] board, String[] words) {
    root = buildTrie(words);
    for (int i = 0; i < board.length; i++) {
      for (int j = 0; j < board[i].length; j++) {
        dfs(board, root, i, j);
      }
    }
    return res;
  }

  private void dfs(char[][] board, TrieNode node, int i, int j) {
    char c = board[i][j];
    if (c == '#') return;
    board[i][j] = '#';  // make move, can't return to prev location
    if (node.children[c - 'a'] != null) node = node.children[c - 'a'];
    if (node.item != "") res.add(node.item);

    // Search in all 4 directions
    if (i > 0) dfs(board, node, i-1, j);
    if (j > 0) dfs(board, node, i, j-1);
    if (i < board.length) dfs(board, node, i+1, j);
    if (j < board[i].length) dfs(board, node, i, j+1);
  }

  private TrieNode buildTrie(String[] words) {
    Trie new_trie = new Trie();
    for (String word : words) {
      new_trie.insert(word);
    }
    return new_trie.root;
  }

}